librz/analysis/arch/tricore/tricore_il.inc

```diff
@@ -0,0 +1,881 @@
+		STOREW(ADD(VARL("EA"), U32(11)), VARG("d8")),
+		STOREW(ADD(VARL("EA"), U32(12)), VARG("a11")),
+		STOREW(ADD(VARL("EA"), U32(13)), VARG("a10")),
+		STOREW(ADD(VARL("EA"), U32(14)), VAR_PSW),
+		STOREW(ADD(VARL("EA"), U32(15)), VAR_PCXI));
+
+	RzILOpEffect *pcxi = NULL;
+	switch (mode) {
+	case CS_MODE_TRICORE_162: {
+		pcxi = SEQ4(set_PCXI_PCPN_v162(ICR_CCPN()),
+			set_PCXI_PCPN_v162(ICR_IE()),
+			set_PCXI_PCPN_v162(U32(1)),
+			setg_bits("PCXI", 0, 20, bits(VARG("FCX"), 0, 20)));
+		break;
+	}
+	case CS_MODE_TRICORE_160: {
+		pcxi = SEQ4(set_PCXI_PCPN_v160(ICR_CCPN()),
+			set_PCXI_PCPN_v160(ICR_IE()),
+			set_PCXI_PCPN_v160(U32(1)),
+			setg_bits("PCXI", 0, 20, bits(VARL("new_FCX"), 0, 20)));
+		break;
+	}
+	default: {
+		RZ_LOG_ERROR("Unknown tricore version: %d\n", mode);
+		return NULL;
+	}
+	}
+
+	RzILOpEffect *fcx = setg_bits("FCX", 0, 20, bits(VARL("new_FCX"), 0, 20));
+
+	RzILOpEffect *jmp = JMP(target);
+
+	ut64 ret = insn->address + 4;
+	RzILOpEffect *a11 = SETG("a11", U32(ret));
+
+	return BRANCH(IS_ZERO(VAR_FCX), trap(FCU),
+		BRANCH(NON_ZERO(PSW_CDE()), cdc_increment(trap(CDO)),
+			SEQ9(cde, tmp_FCX, EA, new_FCX, M, pcxi, fcx, a11,
+				BRANCH(EQ(VARL("new_FCX"), VAR_LCX), trap(FCD), jmp))));
+}
+
+static RzAnalysisLiftedILOp fret() {
+	RzILOpEffect *PC = JMP(SHIFTL0(VARL("A11"), U32(1)));
+	RzILOpEffect *EA = SETL("EA", VARG("a10"));
+	RzILOpEffect *A11 = SETG("a11", LOAD32(VARL("EA")));
+	RzILOpEffect *A10 = SETG("a10", ADD(VARG("a10"), U32(4)));
+	return SEQ5(SETL("A11", VARG("a11")), EA, A11, A10, PC);
+}
+
+static RzILOpPure *is_inf(RzILOpPure *val) {
+	// TODO: check for float Inf
+	return IL_FALSE;
+}
+
+static RzILOpPure *is_nan(RzILOpPure *val) {
+	// TODO: check for float NaN
+	return IL_FALSE;
+}
+
+static RzILOpPure *sign_32bit(RzILOpPure *val) {
+	// TODO: check for signed infinity
+	return IL_FALSE;
+}
+
+static const ut32 ADD_NAN = 0x7fc00001;
+static const ut32 DIV_NAN = 0x7fc00008;
+static const ut32 MUL_NAN = 0x7fc00002;
+static const ut32 SQRT_NAN = 0x7fc00004;
+
+static const ut32 HP_MAX_VALUE = 65504;
+static const ut32 HP_MIN_NORMAL = 1024 * 16;
+
+static const ut32 HP_NEG_INFINITY = 0xfc00;
+static const ut32 HP_POS_INFINITY = 0x7c00;
+
+static const ut32 NEG_INFINITY = 0xff800000;
+static const ut32 POS_INFINITY = 0x7f800000;
+
+/**
+ * \brief Convert single precision to a half precision
+ */
+static RzAnalysisLiftedILOp ftohp(csh handle, cs_insn *insn) {
+	RzILOpPure *a = get_reg(handle, insn, 1);
+	const char *cname = get_regname(handle, insn, 0);
+	RzILOpEffect *inf_eff = BRANCH(sign_32bit(DUP(a)),
+		SETG(cname, U32(HP_NEG_INFINITY)),
+		SETG(cname, U32(HP_POS_INFINITY)));
+
+	RzILOpPure *c = SHIFTL0(BOOL_TO_BV(sign_32bit(DUP(a)), 32), U32(15));
+	c = LOGOR(c, SHIFTL0(U32(0x1f), U32(10)));
+	c = LOGOR(c, SHIFTL0(bits(DUP(a), 21, 2), U32(8)));
+	c = LOGOR(c, bits(DUP(a), 0, 8));
+
+	RzILOpEffect *nan_eff = SETG(cname,
+		ITE(EQ(bits(DUP(c), 0, 10), U32(0)), bits_updated(c, 8, 1, U32(1)), DUP(c)));
```

Comment:
It would be helpful to add comments about the converting algorithm

Reply:
Look here https://www.infineon.com/dgdl/Infineon-AURIX_TC3xx_Architecture_vol2-UserManual-v01_00-EN.pdf?fileId= 5546d46276fb756a01771bc4a6d73b70

Basically the pseudo code is given in the documentation